gk/fib.c

```diff
@@ -1664,3 +1666,329 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+		dentry.len = state.depth;
+		fib = &ltbl->fib_tbl[re4->next_hop];
+		ret = fillup_gk_fib_dump_entry(&dentry, fib);
+		if (ret < 0)
+			goto out;
+
+		lua_getglobal(l, handler);
+		lua_pushlightuserdata(l, &dentry);
+		if (arg == NULL)
+			lua_pushnil(l);
+		else
+			lua_pushlightuserdata(l, arg);
+
+		if (lua_pcall(l, 2, 1, 0) != 0) {
+			lua_error(l);
+			rte_free(dentry.ether_tbl);
+			ret = -1;
+			goto out;
+		}
+
+		rte_free(dentry.ether_tbl);
+
+		index = rte_lpm_rule_iterate(&state, &re4);
+	}
+
+out:
+	return ret;
+}
+
+static int
+list_ipv6_fib_entries(lua_State *l, struct gk_lpm *ltbl,
+	char *handler, void *arg)
+{
+	int index;
+	struct gk_fib *fib;
+	struct rte_lpm6_rule re6;
+	struct rte_lpm6_iterator_state state6;
+	struct gk_fib_dump_entry dentry;
+	int ret = rte_lpm6_iterator_state_init(ltbl->lpm6, NULL, 0, &state6);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize the lpm6 rule iterator state at %s!\n",
+			__func__);
+		goto out;
+	}
+
+	index = rte_lpm6_rule_iterate(&state6, &re6);
+	while (index >= 0) {
+		memset(&dentry, 0, sizeof(dentry));
+		dentry.addr.proto = ETHER_TYPE_IPv6;
+		rte_memcpy(&dentry.addr.ip.v6, re6.ip,
+			sizeof(dentry.addr.ip.v6));
+		dentry.len = re6.depth;
+		fib = &ltbl->fib_tbl6[re6.next_hop];
+		ret = fillup_gk_fib_dump_entry(&dentry, fib);
+		if (ret < 0)
+			goto out;
+
+		lua_getglobal(l, handler);
+		lua_pushlightuserdata(l, &dentry);
+		if (arg == NULL)
+			lua_pushnil(l);
+		else
+			lua_pushlightuserdata(l, arg);
+
+		if (lua_pcall(l, 2, 1, 0) != 0) {
+			lua_error(l);
+			rte_free(dentry.ether_tbl);
+			ret = -1;
+			goto out;
+		}
+
+		rte_free(dentry.ether_tbl);
+
+		index = rte_lpm6_rule_iterate(&state6, &re6);
+	}
+
+out:
+	return ret;
+}
+
+#define CTYPE_STRUCT_GK_CONFIG_PTR "struct gk_config *"
+
+static int
+l_list_fib_entries(lua_State *l)
+{
+	int ret = 0;
+	uint32_t correct_ctypeid;
+
+	uint32_t ctypeid;
+	lua_Integer lpm_cfg;
+	struct gk_config *gk_conf;
+	void *arg;
+	struct gk_lpm *ltbl;
+
+	size_t handler_len;
+	const char *handler_c;
+	char fib_handler[128];
+	memset(fib_handler, 0, sizeof(fib_handler));
+
+	correct_ctypeid = luaL_get_ctypeid(l, CTYPE_STRUCT_GK_CONFIG_PTR);
+
+	/* First argument must be of type CTYPE_STRUCT_GK_CONFIG_PTR. */
+	luaL_checkcdata(l, 1, &ctypeid, CTYPE_STRUCT_GK_CONFIG_PTR);
+	if (ctypeid != correct_ctypeid)
+		luaL_error(l, "Expected `%s' as first argument",
+			CTYPE_STRUCT_GK_CONFIG_PTR);
+
+	/* Second argument must be a number. */
+	luaL_checktype(l, 2, LUA_TNUMBER);
+
+	/* Third argument must be a string. */
+	luaL_checktype(l, 3, LUA_TSTRING);
+
+	gk_conf = *(struct gk_config **)
+		luaL_checkcdata(l, 1, &ctypeid, CTYPE_STRUCT_GK_CONFIG_PTR);
+	lpm_cfg = lua_tointeger(l, 2);
+	handler_c = luaL_checklstring(l, 3, &handler_len);
+	if (handler_len > sizeof(fib_handler))
+		return 0;
+	strncpy(fib_handler, handler_c, handler_len);
+	if (lua_type(l, 4) == LUA_TNIL)
+		arg = NULL;
+	else
+		arg = lua_touserdata(l, 4);
```

Comment:
The handler is a Lua function, so don't assume the `arg` parameter is a user data. Just assume it is a Lua value, that is, any value since we only need to pass it to the handler.